lib/src/executor/host.rs

```diff
@@ -2930,18 +3014,185 @@ impl ExternalOffchainStorageSet {
         HostVm::ReadyToRun(ReadyToRun {
             inner: self.inner,
-            resume_value: None,
+            resume_value: Some(vm::WasmValue::I64(i64::from_ne_bytes(value.to_ne_bytes()))),
         })
     }
 }
 
-impl fmt::Debug for ExternalOffchainStorageSet {
+impl fmt::Debug for OffchainTimestamp {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        f.debug_tuple("ExternalOffchainStorageSet").finish()
+        f.debug_tuple("OffchainTimestamp").finish()
+    }
+}
+
+/// Must return a random seed.
+pub struct OffchainRandomSeed {
+    inner: Box<Inner>,
+
+    /// Function currently being called by the Wasm code. Refers to an index within
+    /// [`VmCommon::registered_functions`]. Guaranteed to be [`FunctionImport::Resolved`̀].
+    calling: usize,
+}
+
+impl OffchainRandomSeed {
+    /// Resumes execution after having set the value.
+    pub fn resume(self, value: [u8; 32]) -> HostVm {
+        let host_fn = match self.inner.common.registered_functions[self.calling] {
+            FunctionImport::Resolved(f) => f,
+            FunctionImport::Unresolved { .. } => unreachable!(),
+        };
+        self.inner
+            .alloc_write_and_return_pointer(host_fn.name(), iter::once(value))
+    }
+}
+
+impl fmt::Debug for OffchainRandomSeed {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_tuple("OffchainRandomSeed").finish()
+    }
+}
+
+/// Must submit an off-chain transaction.
+pub struct OffchainSubmitTransaction {
+    inner: Box<Inner>,
+
+    /// Function currently being called by the Wasm code. Refers to an index within
+    /// [`VmCommon::registered_functions`]. Guaranteed to be [`FunctionImport::Resolved`̀].
+    calling: usize,
+
+    /// Pointer to the transaction whose value must be set. Guaranteed to be in range.
+    tx_ptr: u32,
+
+    /// Size of the transaction whose value must be set. Guaranteed to be in range.
+    tx_size: u32,
+}
+
+impl OffchainSubmitTransaction {
+    pub fn transaction(&'_ self) -> impl AsRef<[u8]> + '_ {
+        self.inner
+            .vm
+            .read_memory(self.tx_ptr, self.tx_size)
+            .unwrap()
+    }
+
+    /// Resumes execution after having set the value.
+    pub fn resume(self, value: (&[u8], usize)) -> HostVm {
+        let host_fn = match self.inner.common.registered_functions[self.calling] {
+            FunctionImport::Resolved(f) => f,
+            FunctionImport::Unresolved { .. } => unreachable!(),
+        };
+
+        let (value, value_total_len) = value;
+        debug_assert_eq!(value.len(), value_total_len);
+
+        let value_len_enc = util::encode_scale_compact_usize(value_total_len);
+        self.inner.alloc_write_and_return_pointer_size(
+            host_fn.name(),
+            iter::once(&[1][..])
+                .chain(iter::once(value_len_enc.as_ref()))
+                .map(either::Left)
+                .chain(iter::once(value).map(either::Right)),
+        )
+    }
```

Comment:
I think that's wrong: https://github.com/paritytech/substrate/blob/689da495a0c0c0c2466fe90a9ea187ce56760f2d/primitives/io/src/lib.rs#L1243

The return value is a `Result<(), ()>`, presumably indicating whether the transaction has successfully been submitted.